server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1201,34 +1204,158 @@ void removeEntitlement(Entitlement entitlement,
+                batch.clear();
+            }
+        }
+        log.info("Modifier entitlements done.");
+
+        if (!batch.isEmpty()) {
+            this.regenerateCertificatesOf(entitlementCurator.batchListModifying(batch), true);
+        }
+
+        Set<Consumer> distinctConsumers = entitlementCurator.getDistinctConsumers(entsToRevoke);
+
+        log.info("Recomputing status for " + distinctConsumers.size() + " consumers.");
+        int i = 1;
+        for (Consumer consumer : distinctConsumers) {
+            if (i++ % 1000 == 0) {
+                consumerCurator.flush();
+            }
+            complianceRules.getStatus(consumer);
+        }
+        consumerCurator.flush();
+
+        log.info("All statuses recomputed.");
+
+        // for each deleted entitlement, create an event
+        for (Entitlement entitlement : entsToRevoke) {
+            Event event = eventFactory.entitlementDeleted(entitlement);
+            sink.queueEvent(event);
+        }
+    }
+
+    /**
+     * Filter the given entitlements so that this method returns only
+     * the entitlements that are part of some stack.
+     * @param entitlements Entitlements to be filtered
+     * @return Entitlements that are stacked
+     */
+    private List<Entitlement> filterStackingEntitlements(List<Entitlement> entitlements) {
+        List<Entitlement> stackingEntitlements = new ArrayList<Entitlement>();
+
+        for (Entitlement ent : entitlements) {
+            Pool pool = ent.getPool();
+
+            if (!"true".equals(pool.getAttributeValue("pool_derived")) &&
+                    pool.hasProductAttribute("stacking_id")) {
+                stackingEntitlements.add(ent);
+            }
+        }
+        return stackingEntitlements;
+    }
+
     @Override
     @Transactional
     public void revokeEntitlement(Entitlement entitlement) {
-        removeEntitlement(entitlement, true);
+        revokeEntitlements(Collections.singletonList(entitlement));
     }
 
     @Override
     @Transactional
     public int revokeAllEntitlements(Consumer consumer) {
-        int count = 0;
-        for (Entitlement e : entitlementCurator.listByConsumer(consumer)) {
-            removeEntitlement(e, false);
-            count++;
-        }
-        // Rerun compliance after removing all entitlements
-        complianceRules.getStatus(consumer);
-        return count;
+        List<Entitlement> entsToDelete = entitlementCurator.listByConsumer(consumer);
+        revokeEntitlements(entsToDelete);
+        return entsToDelete.size();
     }
 
     @Override
     @Transactional
     public int removeAllEntitlements(Consumer consumer) {
-        int count = 0;
-        for (Entitlement e : entitlementCurator.listByConsumer(consumer)) {
-            removeEntitlement(e, false);
-            count++;
-        }
-        return count;
+        List<Entitlement> entsToDelete = entitlementCurator.listByConsumer(consumer);
+        revokeEntitlements(entsToDelete);
```

Comment:
the signature of the method is removeEntitlement(Entitlement e, boolean regenModified). The regenModified attribute when set to false used to skip regeneration and compliance status computation. The regenModified flag seems like a performance optimization, not anything functional.

I can introduce the flag again, but I thought it is not necessary because batch revoke is designed to be fast.